rust/kernel/error.rs

```diff
@@ -5,60 +5,77 @@
 //! C header: [`include/uapi/asm-generic/errno-base.h`](../../../include/uapi/asm-generic/errno-base.h)
 
 use crate::str::CStr;
-use crate::{bindings, c_types};
+use crate::{bindings, c_types, static_assert};
 use alloc::{alloc::AllocError, collections::TryReserveError};
 use core::convert::From;
 use core::fmt;
-use core::num::TryFromIntError;
+use core::num::{NonZeroI32, TryFromIntError};
 use core::str::{self, Utf8Error};
 
 /// Generic integer kernel error.
 ///
 /// The kernel defines a set of integer generic error codes based on C and
 /// POSIX ones. These codes may have a more specific meaning in some contexts.
 ///
+/// We use NonZeroI32 as the inner type because Result<(), Error> can fit in
+/// 32 bit memory, and it takes one register (rather than 2 for i16 or i32)
+/// for returning or argument passing which is fast in many architectures.
+///
 /// # Invariants
 ///
 /// The value is a valid `errno` (i.e. `>= -MAX_ERRNO && < 0`).
 #[derive(Clone, Copy, PartialEq, Eq)]
-pub struct Error(c_types::c_int);
+pub struct Error(NonZeroI32);
+
+/// Converts kernel const to Error.
+/// Note: The value stored in Error is negative
+/// DO NOT use this macro outside of Error const initializations.
+macro_rules! kernel_const_to_error {
+    ($($tt:tt)*) => {
+        {
+            static_assert!(-($($tt)* as i32) >= -(bindings::MAX_ERRNO as i32)
+            && -($($tt)* as i32) < 0);
```

Comment:
This is missing some indentation, or is `rustfmt` forcing it to be like this for some reason?

Reply:
Rustfmt ignores macro bodies as it doesn't know if it should be parsed as expression, pattern, item or type.

Reply:
Thanks! Then I would suggest:

```rust
static_assert!(
    -($($tt)* as i32) >= -(bindings::MAX_ERRNO as i32)
    && -($($tt)* as i32) < 0
);
```